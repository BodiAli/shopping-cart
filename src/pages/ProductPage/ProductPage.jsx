import { useEffect } from "react";
import { useLoaderData, useNavigate, useOutletContext } from "react-router-dom";
import { Splide, SplideSlide } from "@splidejs/react-splide";
import getData from "../../utils/fetchData";
import generatePrice from "../../utils/generatePrice";
import "@splidejs/react-splide/css/sea-green";
import styles from "./ProductPage.module.css";

// Image
import noImage from "../../assets/no-image.svg";

// Icons
import { IoMdAdd, IoMdRemove, IoMdArrowRoundBack } from "react-icons/io";

function ProductPage() {
  const navigate = useNavigate();
  const game = useLoaderData();
  const { arr, handleButtonClick } = useOutletContext();
  const [cart] = arr;

  useEffect(() => {
    document.title = `${game.title} | GameVault`;

    return () => {
      document.title = "GameVault";
    };
  }, [game.title]);

  const price = generatePrice();
  game.price = price;
  game.originalPrice = price;
  game.quantity = 1;

  function handleBackButton() {
    navigate(-1);
  }
  return (
    <main className={styles.main}>
      <button onClick={handleBackButton} className={styles.backButton}>
        <IoMdArrowRoundBack /> Back
      </button>
      <p className={styles.screenshotsP}>Game Screenshots</p>
      <Splide
        data-testid="carousel"
        tag="section"
        options={{
          type: "loop",
          drag: true,
          autoplay: true,
          interval: 5000,
          arrows: false,
        }}
        className={styles.carousel}
      >
        {game.sample_screenshots.length === 0 ? (
          <img className={styles.noImage} src={noImage} alt="no image" />
        ) : (
          game.sample_screenshots.map((screenshot, index) => {
            return (
              <SplideSlide key={index}>
                <div className={styles.screenshotsContainer}>
                  <img
                    className={styles.screenshot}
                    src={screenshot.thumbnail_image ? screenshot.thumbnail_image : noImage}
                    alt="game screenshot"
                  />
                </div>
              </SplideSlide>
            );
          })
        )}
      </Splide>

      <section className={styles.gameInfo}>
        <div className={styles.infoImgContainer}>
          <img
            src={game.sample_cover ? game.sample_cover.thumbnail_image : noImage}
            alt={`${game.title} game poster`}
          />
        </div>
        <div className={styles.container}>
          <div className={styles.titlesContainer}>
            <div className={styles.titleContainer}>
              <p>Title:</p>
              <h2 className={styles.title}>{game.title}</h2>
            </div>
            <div className={styles.alternateTitlesContainer}>
              <p>Alternate Titles:</p>
              <ul className={styles.alternateTitles}>
                {game.alternate_titles.length === 0 ? (
                  <li>No alternate titles</li>
                ) : (
                  game.alternate_titles.map((title, index) => {
                    return (
                      <div key={index}>
                        <li>{title.title}</li>
                        <ul>
                          <li>{title.description}</li>
                        </ul>
                      </div>
                    );
                  })
                )}
              </ul>
            </div>
          </div>
          <div className={styles.genresContainer}>
            <p>Genres:</p>
            <ul className={styles.genres}>
              {game.genres.map((genre, index) => {
                return (
                  <div key={index}>
                    <li>{genre.genre_category}</li>
                    <ul>
                      <li>{genre.genre_name}</li>
                    </ul>
                  </div>
                );
              })}
            </ul>
          </div>
          <div className={styles.platformsContainer}>
            <p>Platforms:</p>
            <ul className={styles.platforms}>
              {game.platforms.map((platform, index) => {
                return (
                  <div key={index}>
                    <li>{platform.platform_name}</li>
                    <ul>
                      <li>Release date: {platform.first_release_date}</li>
                    </ul>
                  </div>
                );
              })}
            </ul>
          </div>

          <div className={styles.buttonContainer}>
            <p>
              Price: <span>{game.originalPrice}</span>
            </p>
            <button onClick={(e) => handleButtonClick(e, game)}>
              {cart.isTheProductPresent(game) ? (
                <>
                  <IoMdRemove color="#c00000" /> Remove from Cart
                </>
              ) : (
                <>
                  <IoMdAdd color="#01f5b8" /> Add to Cart
                </>
              )}
            </button>
          </div>
        </div>
        <div
          className={styles.descriptionContainer}
          dangerouslySetInnerHTML={{ __html: game.description }}
        ></div>
      </section>
    </main>
  );
}

async function loader({ params }) {
  try {
    const game = await getData(`?id=${params.id}`);
    if (!game.games || game.games.length === 0) {
      throw new Error("Data not found");
    }

    return game.games[0];
  } catch (error) {
    if (error.message === "Bad Request") {
      throw new Error("Data not found");
    } else {
      throw new Error(error);
    }
  }
}

ProductPage.loader = loader;

export default ProductPage;
