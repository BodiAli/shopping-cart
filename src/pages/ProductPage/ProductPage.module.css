.main {
  background-color: var(--green);
  padding: 2rem;
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.screenshotsP {
  text-align: center;
  font-size: 2.5rem;
  font-family: "Merriweather", Times, serif;
}

.carousel {
  margin: 0 200px;
  padding: 0;
}

.screenshot {
  width: 100%;
  height: 400px;
  object-fit: cover;
  object-position: center;
  border-radius: 5px;
}

.noImage {
  width: 100%;
  height: 400px;
  object-fit: contain;
}

.gameInfo {
  display: grid;
  grid-template-columns: 1fr 4fr;
  box-shadow: 2.8px 2.8px 10px rgba(0, 0, 0, 0.775);
}

.infoImgContainer {
  background-color: var(--primary);
  display: flex;
}

.infoImgContainer > img {
  align-self: center;
}

.gameInfo > .container {
  display: flex;
  flex-direction: column;
  gap: 20px;
  padding: 0.5rem;
  background-color: var(--gray);
}

.buttonContainer {
  margin-top: 20px;
  align-self: center;
}

.buttonContainer > button {
  font-family: "Merriweather", Times, serif;
  font-size: 1.3rem;
  font-weight: bold;
  background: var(--darkerGray);
  border: none;
  border-radius: 5px;
  padding: 0.8rem 1.6rem;
  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 3px;
}

.buttonContainer > button:hover {
  filter: brightness(1.2);
}

.buttonContainer > button:active {
  filter: brightness(0.9);
}

.titlesContainer {
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.titleContainer {
  font-family: "Merriweather", Times, serif;
  font-size: 2.2rem;
  display: flex;
  gap: 5px;
  font-weight: 600;
}

.alternateTitlesContainer {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.alternateTitlesContainer > p {
  font-size: 1.5rem;
}

.genresContainer {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.genresContainer > p {
  font-size: 1.5rem;
}

.platformsContainer {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.platformsContainer > p {
  font-size: 1.5rem;
}

.gameInfo > .descriptionContainer {
  grid-column: 1 / 3;
  background-color: var(--darkerGreen);
  padding: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.alternateTitles {
  list-style: disc;
  list-style-position: inside;
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  font-size: 1.1rem;
}

.alternateTitles > div {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.alternateTitles ul {
  padding-inline-start: 10px;
  list-style-type: circle;
  list-style-position: inside;
}

.genres {
  list-style: disc;
  list-style-position: inside;
  display: flex;
  flex-wrap: wrap;
  gap: 15px;
  font-size: 1.1rem;
}

.genres > div {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.genres ul {
  padding-inline-start: 10px;
  list-style-type: circle;
  list-style-position: inside;
}

.platforms {
  list-style: disc;
  list-style-position: inside;
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  font-size: 1.1rem;
}

.platforms > div {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.platforms ul {
  padding-inline-start: 10px;
  list-style-type: circle;
  list-style-position: inside;
}
