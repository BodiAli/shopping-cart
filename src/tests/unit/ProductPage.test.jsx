import { render, screen } from "@testing-library/react";
import ProductPage from "../../pages/ProductPage/ProductPage";
import { vi } from "vitest";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import Cart from "../../utils/cart";

const cart = new Cart();

const handleButtonClick = vi.fn();
const games = JSON.stringify({
  games: [
    {
      alternate_titles: [
        {
          title: "block game",
          description: "another title",
        },
      ],

      title: "Minecraft game",
      sample_cover: {
        thumbnail_image: "image.jpg",
      },
      description: "<p>A game where you build stuff</p>",
      genres: [
        {
          genre_category: "Basic genres",
          genre_name: "Adventure",
        },
      ],

      platforms: [
        {
          platform_name: "Windows",
          first_release_date: "2012",
        },
      ],
      originalPrice: "50.50",
      sample_screenshots: [
        {
          thumbnail_image: "screenshot.jpg",
        },
      ],
    },
  ],
});
const MOCK_GAME = {
  contents: games,
};

// eslint-disable-next-line no-undef
global.fetch = vi.fn(() =>
  Promise.resolve({
    ok: true,
    json: vi.fn(() => Promise.resolve(MOCK_GAME)),
  })
);

// Mocking useOutletContext to return an array because it returns null and causes an error
vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useOutletContext: vi.fn(() => {
      return { arr: [cart], handleButtonClick };
    }),
  };
});

const router = createMemoryRouter([
  {
    path: "/",
    element: <ProductPage />,
    loader: ProductPage.loader,
  },
]);

const wrapper = () => <RouterProvider router={router} />;

describe("Product page component", () => {
  test("should render correct elements", () => {
    const { container } = render(<ProductPage />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("should change document title", () => {
    render(<ProductPage />, { wrapper });
    expect(document.title).toBe("Minecraft game | GameVault");
  });

  test("should render a back button", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByRole("button", { name: /back/i })).toBeInTheDocument();
  });

  test("should render game screenshots", () => {
    render(<ProductPage />, { wrapper });
    const carousel = screen.getByTestId("carousel");
    const images = screen.getAllByAltText("game screenshot");
    images.forEach((image) => {
      expect(carousel).toContainElement(image);
    });
    expect(screen.getByText(/game screenshots/i)).toBeInTheDocument();
  });

  test("should render game cover image", () => {
    render(<ProductPage />, { wrapper });
    const image = screen.getByAltText(/minecraft game game poster/i);
    expect(image.src).toContain("image.jpg");
  });

  test("should render game title", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent("Minecraft game");
  });

  test("should render game alternate titles", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByText(/alternate titles:/i)).toBeInTheDocument();
    expect(screen.getByText(/block game/i)).toBeInTheDocument();
    expect(screen.getByText(/another title/i)).toBeInTheDocument();
  });

  test("should render game genres", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByText(/genres:/i)).toBeInTheDocument();
    expect(screen.getByText(/basic genres/i)).toBeInTheDocument();
    expect(screen.getByText(/adventure/i)).toBeInTheDocument();
  });

  test("should render game platforms", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByText(/platforms:/i)).toBeInTheDocument();
    expect(screen.getByText(/windows/i)).toBeInTheDocument();
    expect(screen.getByText(/2012/i)).toBeInTheDocument();
  });

  test("should render game description", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByText(/a game where you build stuff/i)).toBeInTheDocument();
  });

  test("should render game price", () => {
    render(<ProductPage />, { wrapper });
    expect(screen.getByText(/price:/i)).toBeInTheDocument();
    expect(screen.getByText(/50.50/i)).toBeInTheDocument();
  });
});
