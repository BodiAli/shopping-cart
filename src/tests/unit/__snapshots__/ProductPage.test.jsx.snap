// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product page component > should render correct elements 1`] = `
<div>
  <main
    class="_main_973786"
  >
    <button
      class="_backButton_973786"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M401.4 224h-214l83-79.4c11.9-12.5 11.9-32.7 0-45.2s-31.2-12.5-43.2 0L89 233.4c-6 5.8-9 13.7-9 22.4v.4c0 8.7 3 16.6 9 22.4l138.1 134c12 12.5 31.3 12.5 43.2 0 11.9-12.5 11.9-32.7 0-45.2l-83-79.4h214c16.9 0 30.6-14.3 30.6-32 .1-18-13.6-32-30.5-32z"
        />
      </svg>
       Back
    </button>
    <p
      class="_screenshotsP_973786"
    >
      Game Screenshots
    </p>
    <section
      aria-roledescription="carousel"
      class="splide _carousel_973786 splide--loop splide--ltr splide--draggable is-active is-initialized"
      data-testid="carousel"
      id="splide01"
    >
      <div
        aria-atomic="true"
        aria-live="off"
        class="splide__track splide__track--loop splide__track--ltr splide__track--draggable"
        id="splide01-track"
        style="padding-left: 0px; padding-right: 0px;"
      >
        <ul
          class="splide__list"
          id="splide01-list"
          role="presentation"
          style="transform: translateX(0px);"
        >
          <li
            aria-label="1 of 1"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-active is-visible"
            id="splide01-clone01"
            role="tabpanel"
          >
            <div
              class="_screenshotsContainer_973786"
            >
              <img
                alt="game screenshot"
                class="_screenshot_973786"
                src="screenshot.jpg"
              />
            </div>
          </li>
          <li
            aria-label="1 of 1"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-active is-visible is-prev"
            id="splide01-clone02"
            role="tabpanel"
          >
            <div
              class="_screenshotsContainer_973786"
            >
              <img
                alt="game screenshot"
                class="_screenshot_973786"
                src="screenshot.jpg"
              />
            </div>
          </li>
          <li
            aria-label="1 of 1"
            aria-roledescription="slide"
            class="splide__slide is-active is-visible"
            id="splide01-slide01"
            role="tabpanel"
          >
            <div
              class="_screenshotsContainer_973786"
            >
              <img
                alt="game screenshot"
                class="_screenshot_973786"
                src="screenshot.jpg"
              />
            </div>
          </li>
          <li
            aria-label="1 of 1"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-active is-visible is-next"
            id="splide01-clone03"
            role="tabpanel"
          >
            <div
              class="_screenshotsContainer_973786"
            >
              <img
                alt="game screenshot"
                class="_screenshot_973786"
                src="screenshot.jpg"
              />
            </div>
          </li>
          <li
            aria-label="1 of 1"
            aria-roledescription="slide"
            class="splide__slide splide__slide--clone is-active is-visible"
            id="splide01-clone04"
            role="tabpanel"
          >
            <div
              class="_screenshotsContainer_973786"
            >
              <img
                alt="game screenshot"
                class="_screenshot_973786"
                src="screenshot.jpg"
              />
            </div>
          </li>
        </ul>
      </div>
      <ul
        aria-label="Select a slide to show"
        class="splide__pagination splide__pagination--ltr"
        role="tablist"
      >
        <li
          role="presentation"
        >
          <button
            aria-controls="splide01-slide01"
            aria-label="Go to slide 1"
            aria-selected="true"
            class="splide__pagination__page is-active"
            role="tab"
            type="button"
          />
        </li>
      </ul>
    </section>
    <section
      class="_gameInfo_973786"
    >
      <div
        class="_infoImgContainer_973786"
      >
        <img
          alt="Minecraft game game poster"
          src="image.jpg"
        />
      </div>
      <div
        class="_container_973786"
      >
        <div
          class="_titlesContainer_973786"
        >
          <div
            class="_titleContainer_973786"
          >
            <p>
              Title:
            </p>
            <h2
              class="_title_973786"
            >
              Minecraft game
            </h2>
          </div>
          <div
            class="_alternateTitlesContainer_973786"
          >
            <p>
              Alternate Titles:
            </p>
            <ul
              class="_alternateTitles_973786"
            >
              <div>
                <li>
                  block game
                </li>
                <ul>
                  <li>
                    another title
                  </li>
                </ul>
              </div>
            </ul>
          </div>
        </div>
        <div
          class="_genresContainer_973786"
        >
          <p>
            Genres:
          </p>
          <ul
            class="_genres_973786"
          >
            <div>
              <li>
                Basic genres
              </li>
              <ul>
                <li>
                  Adventure
                </li>
              </ul>
            </div>
          </ul>
        </div>
        <div
          class="_platformsContainer_973786"
        >
          <p>
            Platforms:
          </p>
          <ul
            class="_platforms_973786"
          >
            <div>
              <li>
                Windows
              </li>
              <ul>
                <li>
                  Release date: 
                  2012
                </li>
              </ul>
            </div>
          </ul>
        </div>
        <div
          class="_buttonContainer_973786"
        >
          <p>
            Price: 
            <span>
              50.50
            </span>
          </p>
          <button>
            <svg
              color="#01f5b8"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(1, 245, 184);"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 277.333H277.333V416h-42.666V277.333H96v-42.666h138.667V96h42.666v138.667H416v42.666z"
              />
            </svg>
             Add to Cart
          </button>
        </div>
      </div>
      <div
        class="_descriptionContainer_973786"
      >
        <p>
          A game where you build stuff
        </p>
      </div>
    </section>
  </main>
</div>
`;
